designs/v1beta1-api.md

```diff
@@ -0,0 +1,341 @@
+    2. Neutral behavior-based fields for configuring Karpenter’s scheduling and deprovisioning decision-making
+2. `NodeClass`
+    1. Cloudprovider-specific Node configuration-based fields that affect launch and bootstrap process for that Node including: configuring startup scripts (including kubelet configuration), volume mappings, metadata settings, etc.
+    2. Cloudprovider-specific behavior-based fields for configuring Karpenter’s scheduling and deprovisioning decision-making (e.g. interruption-based disruption, allocation strategy)
+3. `NodeClaim`
+    1. A Karpenter management object that fully manages the lifecycle of a single node including: configuring and launching the node, monitoring the node health (including disruption conditions), and handling the deprovisioning and termination of the node
+
+### `karpenter.sh/NodePool`
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec:
+  template:
+    metadata:
+      labels:
+        billing-team: my-team
+      annotations:
+        example.com/owner: "my-team"
+    spec:
+      nodeClass:
+        name: default
+        kind: NodeClass
+        apiVersion: compute.k8s.aws/v1beta1
+      taints:
+        - key: example.com/special-taint
+          effect: NoSchedule
+      startupTaints:
+        - key: example.com/another-taint
+          effect: NoSchedule
+      requirements:
+        - key: "karpenter.k8s.aws/instance-category"
+          operator: In
+          values: ["c", "m", "r"]
+      resources:
+        requests:
+          cpu: "1"
+          memory: "100Mi"
+      kubeletConfiguration:
+        clusterDNS: ["10.0.1.100"]
+        containerRuntime: containerd
+        systemReserved:
+          cpu: 100m
+          memory: 100Mi
+          ephemeral-storage: 1Gi
+        kubeReserved:
+          cpu: 200m
+          memory: 100Mi
+          ephemeral-storage: 3Gi
+        evictionHard:
+          memory.available: 5%
+          nodefs.available: 10%
+          nodefs.inodesFree: 10%
+        evictionSoft:
+          memory.available: 500Mi
+          nodefs.available: 15%
+          nodefs.inodesFree: 15%
+        evictionSoftGracePeriod:
+          memory.available: 1m
+          nodefs.available: 1m30s
+          nodefs.inodesFree: 2m
+        evictionMaxPodGracePeriod: 60
+        imageGCHighThresholdPercent: 85
+        imageGCLowThresholdPercent: 80
+        cpuCFSQuota: true
+        podsPerCore: 2
+        maxPods: 20
+  deprovisioning:
+    consolidationTTL: 10m
+    consolidationPolicy: WhenEmpty | WhenUnderutilized
+    expirationTTL: 30d
+    uninitializedTTL: 20m
+  weight: 10
+  limits:
+    cpu: "1000"
+    memory: 1000Gi
+status:
+  resources:
+     cpu: "2"
+     memory: "100Mi"
+     ephemeral-storage: "100Gi"
+```
+
+### `compute.k8s.aws/NodeClass`
+
+```
+apiVersion: compute.k8s.aws/v1beta1
+kind: NodeClass
+metadata:
+  name: default
+spec:
+  amiFamily: AL2
+  amiSelector:
```

Comment:
This isn't really a selector (not in the Kubernetes sense). Can we call it a filter instead?

Reply:
I'm curious why you say this isn't a selector. Agreed that isn't a label selector or a field selector but (from what I can tell), there is prior art with label selectors with `matchExpressions` that allows you to select on multiple ORed things.

Reply:
It's not a selector because Kubernetes selectors have certain qualities (eg: always a logical AND), that this doesn't align with.

If we call it a selector, people may assume it works like other selectors - but it may not.

Reply:
These selectors are absolutely a logical AND. All selector terms are applied to filter down to a set of selected objects. For objects that can use multiple (e.g. security groups), we use them all. For objects that are singleton (e.g. sunset/ami) we pick one. 

This is no different than node selector requirements that use an AND to filter down to a known list, and then we pick one (e.g. instance type, zone).

Reply:
I think there's some confusion here around what is getting ANDed and what is getting ORed. I've changed the name to `amiSelectorTerms` since I agree that this better aligns with `nodeSelectorTerms`. Each requirement in a given term will be ANDed together. For example if we have an `amiSelectorTerm`

```yaml
amiSelectorTerms:
- tags:
     key: value
   owner: amazon
```

This implies that we will look across all AMIs that have `amazon` as their owner AND also have the tag `key: value`. Each of these terms is ORed together to create the full list of available AMIs.

Reply:
@jonathan-innis set me straight. This is a list (ORed) of terms (ANDed), identical to node selector.

Reply:
Correct.

Reply:
I'm still a bit dubious about

```
amiSelectorTerms:
- tags:
     key: value
   owner: amazon
```
but since this isn't Karpenter core, I really don't think it should block the beta.